lib/readline.js

```diff
@@ -327,6 +329,24 @@ Interface.prototype._addHistory = function() {
   return this.history[0];
 };
 
+// called when a multiline is seen by the repl
+Interface.prototype.undohistory = function() {
+  this.history && this.history.shift();
+};
```

Comment:
👆 I don't think `this.history &&` is needed. All other places in this module assumes `this.history` exists without guards _(it's initialized with an empty array)_.

Reply:
@jdalton  Yes initially, I didn't add that check, but the below test case failed:

```
Path: parallel/test-repl-tab-complete
assert.js:739
    throw newErr;
    ^

AssertionError [ERR_ASSERTION]: ifError got unwanted exception: Cannot read property 'shift' of undefined
    at Domain.<anonymous> (/Users/anto/programs/node/node/test/parallel/test-repl-tab-complete.js:57:10)
    at Domain.emit (events.js:187:15)
    at Domain.EventEmitter.emit (domain.js:442:20)
    at Domain._errorHandler (domain.js:224:23)
    at Object.setUncaughtExceptionCaptureCallback (domain.js:140:29)
    at process._fatalException (internal/bootstrap/node.js:535:31)
    at REPLServer.Interface.undohistory (readline.js:334:16)
    at REPLServer.displayPrompt (repl.js:895:45)
    at finish (repl.js:667:16)
    at finishExecution (repl.js:309:7)
    at REPLServer.defaultEval (repl.js:401:7)
    at bound (domain.js:396:14)
    at REPLServer.runBound [as eval] (domain.js:409:12)
    at REPLServer.onLine (repl.js:644:10)
    at REPLServer.emit (events.js:182:13)
    at REPLServer.EventEmitter.emit (domain.js:442:20)
Command: out/Release/node /Users/anto/programs/node/node/test/parallel/test-repl-tab-complete.js
```

Reply:
I think that may be a sign of another issue _(maybe it's being used in an unbound cherry-picked method kind of way)_. I think you should look into that instead of adding a one-or-two-off guard here.

Reply:
@jdalton: Sure, will have a look at them.

Reply:
@jdalton I figured out that the above failing test cases where not `isTTY`, hence we are not setting up the `history` in our readline `Interface`. I have added a check only if its a `terminal` undo the operation. Can you have a look at them now? I will add test cases for the same and update the PR in some time.